libs/constrained-generators/src/Constrained/Spec/Generics.hs

```diff
@@ -414,11 +422,126 @@ instance
   where
   prod_ (a :> as) = app pairFn a (prod_ as)
 
+type family At n as where
+  At 0 (a : as) = a
+  At n (a : as) = At (n - 1) as
+
+class Select fn n as where
+  select_ :: Term fn (ProdOver as) -> Term fn (At n as)
+
+instance Select fn 0 (a : '[]) where
+  select_ = id
+
+instance (HasSpec fn a, HasSpec fn (ProdOver (a' : as))) => Select fn 0 (a : a' : as) where
+  select_ = app fstFn
+
+instance
+  {-# OVERLAPPABLE #-}
+  ( HasSpec fn a
+  , HasSpec fn (ProdOver (a' : as))
+  , At (n - 1) (a' : as) ~ At n (a : a' : as)
+  , Select fn (n - 1) (a' : as)
+  ) =>
+  Select fn n (a : a' : as)
+  where
+  select_ = select_ @fn @(n - 1) @(a' : as) . app sndFn
+
+class IsConstrOf (c :: Symbol) b sop where
+  mkCases ::
+    (HasSpec fn b, All (HasSpec fn) (Cases (SOP sop))) =>
+    (forall a. Term fn a -> Pred fn) ->
+    (Term fn b -> Pred fn) ->
+    List (Binder fn) (Cases (SOP sop))
```

Comment:
OK, so we are now doing something new for Constructors?
Maybe some explanation.
Was the addition of this  performance issue?

Reply:
It's an expressiveness thing. All in "client code". Basically you want convenient functions for saying "if its this constructor, these constraints apply" and "it has to be this constructor" without having to write the big caseOf statements